domain/keymanager/state/state.go

```diff
@@ -236,40 +457,68 @@ WHERE user_id = $userId.user_id
 }
 
 // GetPublicKeysDataForUser is responsible for returning all of the public keys
-// raw data for the user id in this model. If the user does not exist no error
-// is returned.
+// raw data for the user uuid on a given model.
+// The following error can be expected:
+// - accesserrors.UserNotFound if the user does not exist.
+// - modelerrors.NotFound if the model does not exist.
 func (s *State) GetPublicKeysDataForUser(
 	ctx context.Context,
-	id user.UUID,
+	modelUUID model.UUID,
+	userUUID user.UUID,
 ) ([]string, error) {
 	db, err := s.DB()
 	if err != nil {
-		return nil, errors.Trace(err)
+		return nil, err
 	}
 
-	userId := userId{id.String()}
+	userUUIDVal := userUUIDValue{userUUID.String()}
+	modelUUIDVal := modelUUIDValue{modelUUID.String()}
 
 	stmt, err := s.Prepare(`
 SELECT (public_key) AS (&publicKeyData.*)
-FROM user_public_ssh_key
-WHERE user_id = $userId.user_id
-`, userId, publicKeyData{})
-
+FROM user_public_ssh_key AS upsk
+INNER JOIN model_authorized_keys AS m ON upsk.id = m.user_public_ssh_key_id
+WHERE user_uuid = $userUUIDValue.user_uuid
+AND model_uuid = $modelUUIDValue.model_uuid
+`, userUUIDVal, modelUUIDVal, publicKeyData{})
 	if err != nil {
-		return nil, errors.Trace(err)
+		return nil, errors.Errorf(
+			"preparing user %q public keys data statement on model %q: %w",
+			userUUID, modelUUID, err,
+		)
 	}
 
 	publicKeys := []publicKeyData{}
 	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
-		return tx.Query(ctx, stmt, userId).GetAll(&publicKeys)
+		err := s.checkUserExists(ctx, userUUID, tx)
+		if err != nil {
+			return errors.Errorf(
+				"getting public keys data for user %q on model %q: %w",
+				userUUID, modelUUID, err,
+			)
+		}
+
+		err = s.checkModelExists(ctx, modelUUID, tx)
+		if err != nil {
+			return errors.Errorf(
+				"getting public keys data for user %q on model %q: %w",
+				userUUID, modelUUID, err,
+			)
+		}
+
+		err = tx.Query(ctx, stmt, userUUIDVal, modelUUIDVal).GetAll(&publicKeys)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
```

Comment:
Do we return nothing if no key is found?

Reply:
Can you define nothing? The code down below will initialise a zero length slice and return that if nothing is found.

Reply:
Sorry, I mean why we ignore the not found error.

Reply:
There is no not found error in this context or do you mean the no rows error? If it is the latter we ignore it because it just means there is no keys to list and that is not an error within Juju.